parachain/pallets/control/src/lib.rs

```diff
@@ -311,25 +389,49 @@ pub mod pallet {
 
 		pub fn convert_location(
 			mut location: MultiLocation,
-		) -> Result<(H256, Option<ParaId>, MultiLocation), DispatchError> {
+		) -> Result<(H256, ParaId, MultiLocation), DispatchError> {
 			// Normalize all locations relative to the relay chain.
 			let relay_location = T::RelayLocation::get();
 			location
 				.reanchor(&relay_location, T::UniversalLocation::get())
 				.map_err(|_| Error::<T>::LocationReanchorFailed)?;
 
-			// Only allow Parachain as origin location
-			let para_id = match location {
-				MultiLocation { parents: 0, interior: X1(Parachain(index)) } =>
-					Some((index).into()),
+			let para_id = match location.interior.first() {
+				Some(Parachain(index)) => Some((*index).into()),
 				_ => None,
-			};
+			}
+			.ok_or(Error::<T>::LocationToParaIdConversionFailed)?;
 
 			// Hash the location to produce an agent id
 			let agent_id = T::AgentHashedDescription::convert_location(&location)
 				.ok_or(Error::<T>::LocationToAgentIdConversionFailed)?;
 
 			Ok((agent_id, para_id, location))
 		}
+
+		pub fn account_id() -> T::AccountId {
+			PalletId(*b"snow/ctl").into_account_truncating()
+		}
+
+		pub fn reserve_deposit(
+			payer: T::AccountId,
+			reserve_deposit: Option<BalanceOf<T>>,
+		) -> DispatchResult {
+			if reserve_deposit.is_some() {
+				T::Token::transfer(
+					&payer,
+					&Self::account_id(),
+					reserve_deposit.unwrap(),
+					Preservation::Preserve,
+				)?;
+			}
+			Ok(())
+		}
+
+		pub fn agent_account_id(agent_id: H256) -> Result<T::AccountId, DispatchError> {
+			let agent_owner = T::AccountId::decode(&mut &agent_id.as_fixed_bytes()[..])
+				.map_err(|_| Error::<T>::AgentIdToAccountConversionFailed)?;
+			Ok(agent_owner)
+		}
```

Comment:
There is already a converter in Polkadot that we can reuse here:

https://github.com/paritytech/polkadot-sdk/blob/73915d3459a0723e2ca80a5ca5d0a11a3736b18d/polkadot/xcm/xcm-builder/src/location_conversion.rs#L120

So you can add a type to the config like: `SovereignAccountOf: ConvertLocation<AccountId>`

Reply:
Exactly, essentially [AgentHashedDescription](https://github.com/Snowfork/snowbridge/blob/ffa44a7cce7388d5bb32211922c12f2626d22870/parachain/primitives/router/src/outbound/mod.rs#L18) is just an implementation of trait `ConvertLocation` though we provide `H256` here.

And I've added another [AgentAccountDescription](https://github.com/Snowfork/snowbridge/blob/ffa44a7cce7388d5bb32211922c12f2626d22870/parachain/primitives/router/src/outbound/mod.rs#L20) convert to `AccountId32` to make it less confusing.

But actually just the same as before and identical as the `agent_id`,  check the test [here](https://github.com/Snowfork/snowbridge/blob/ffa44a7cce7388d5bb32211922c12f2626d22870/parachain/pallets/control/src/tests.rs#L42) FYI.

Reply:
I feel differently. I would rather have two separate converters defined in the config:

```rust
    type AgentIdOf: ConvertLocation<H256>;
    type SovereignAccountOf: ConvertLocation<AccountId>
```

This makes it clear that the converters are being used for different purposes.

Reply:
Make sense! some revamp in https://github.com/Snowfork/snowbridge/pull/940/commits/6cc82e5ed1bd849694e863e24f5674ba6c87cbbd please check.

Reply:
Looks much better!